src/components/ColumnPicker/useColumns.ts

```diff
@@ -0,0 +1,164 @@
+          ),
+      );
+    }
+  }, [columnCategories, columnPickerSettings, dynamicColumns]);
+
+  const initialColumnsState = React.useMemo(
+    () =>
+      [...staticColumns, ...(dynamicColumns ?? [])].filter(
+        column => selectedColumns.includes(column.id) || column.id === "empty",
+      ),
+    [dynamicColumns, staticColumns, selectedColumns],
+  );
+
+  const [recentlyAddedColumn, setRecentlyAddedColumn] = React.useState<
+    string | null
+  >(null);
+
+  const [visibleColumns, setVisibleColumns] =
+    useStateFromProps(initialColumnsState);
+
+  const onMove = React.useCallback(
+    (startIndex: number, endIndex: number): void => {
+      // When empty column prevent to rearrange it order
+      if (visibleColumns[0]?.id === "empty") {
+        if (startIndex === 0) {
+          return;
+        }
+
+        // Keep empty column always at beginning
+        if (endIndex === 0) {
+          return setVisibleColumns(old =>
+            addAtIndex(
+              old[startIndex],
+              removeAtIndex(old, startIndex),
+              endIndex + 1,
+            ),
+          );
+        }
+      }
+
+      setVisibleColumns(old =>
+        addAtIndex(old[startIndex], removeAtIndex(old, startIndex), endIndex),
+      );
+    },
+    [visibleColumns, setVisibleColumns],
+  );
+
+  const onResize = React.useCallback(
+    (column: GridColumn, newSize: number) => {
+      if (column.id === "empty") {
+        return;
+      }
+      return setVisibleColumns(prevColumns =>
+        prevColumns.map(prevColumn =>
+          prevColumn.id === column.id
+            ? { ...prevColumn, width: newSize }
+            : prevColumn,
+        ),
+      );
+    },
+    [setVisibleColumns],
+  );
+
+  const onChange = (columns: string[]) => {
+    // Recently added is used by datagrid to auto-scroll to the column
+    setRecentlyAddedColumn(difference(columns, selectedColumns)[0]);
+    // Saves in LS
+    onSave(columns);
+  };
+
+  // Should be used only for special cases
+  const onCustomUpdateVisible = setVisibleColumns;
+
+  const onDynamicColumnSelect = (columns: string[]) => {
+    if (typeof setDynamicColumnSettings !== "function") {
+      return;
+    }
+    setDynamicColumnSettings(columns);
+    updateDynamicColumns(prevDynamicColumns =>
+      uniqBy(
+        [
+          ...(prevDynamicColumns ?? []),
+          ...columnCategories
+            .flatMap(category => category.availableNodes)
+            .filter(x => columns.includes(x.id)),
+        ],
+        "id",
+      )
+        .sort((a, b) => columns.indexOf(a.id) - columns.indexOf(b.id))
+        .filter(x => columns.includes(x.id)),
+    );
+  };
+
+  return {
+    handlers: {
+      onMove,
+      onResize,
+      onChange,
+      onDynamicColumnSelect,
+      onCustomUpdateVisible,
+    },
+    visibleColumns,
+    staticColumns,
+    dynamicColumns,
+    selectedColumns,
+    columnCategories,
+    columnPickerSettings,
+    recentlyAddedColumn,
+  };
+};
```

Comment:
is this file yours? or you copied that? if so, please do very little refactoring if you can, I get that if it's not yours